### PR TITLE
tests for abstract densities

### PR DESCRIPTION
**Pull request by sthayashi**

### CODE REVIEW

**Comment by oschulz** [comment]
If you're done with this, can you remove `[WIP]` from the title of the PR?

**Comment by sthayashi** [comment]
I will be writing more tests for this, unless it's more desirable to write tests for something else. 

https://github.com/bat/BAT.jl/pull/184 should be ready for review

**Comment by oschulz** [comment]
Sure, that's perfectly fine!

**Comment by sthayashi** [comment]
It seems the function `map_const_shapes(f::Function, shape::ConstValueShape)` is only used for statistics of a `DensitySampleVector`. Should this be tested in `test_abstract_density.jl`, or is it more appropriate with the other `density_sample.jl` tests?

**Comment by oschulz** [comment]
`map_const_shapes` will move into ValueShapes, doesn't need a test in BAT

**Comment by sthayashi** [comment]
Does it make sense that this works?

```
nt_dist = NamedTupleDist(a=Normal())
logpdf(nt_dist, [0])
logpdf(nt_dist, [0, 1])
logpdf(nt_dist, rand(1000))
```
It looks like it takes the first element to evaluate, i.e. `logpdf(Normal(), 0)` for all of them. Just ran into this as I was trying to test `_density_eval_error` on line 163, and am wondering if it should also throw an error when the the number of elements doesn't match the `AbstractDensity` when using `logvalof(density::AbstractDensity, v::Any)`

**Comment by oschulz** [comment]
> Does it make sense that this works

That should indeed result in an error.

**Comment by sthayashi** [comment]
Just need to write one more test for 

https://github.com/bat/BAT.jl/blob/a934f6845111aba431e444fd11be32676ff5543c/src/densities/abstract_density.jl#L195

I think this is just used for making the raised error text more readable? Is it good enough to do something simple like

`@test_throws ErrorException logvalgradof(some_dist::AbstractDensity, v::Vector)`.

I don't think there is a way to more directly test this?

The relevant chain of code which ends up calling this is
https://github.com/bat/BAT.jl/blob/a934f6845111aba431e444fd11be32676ff5543c/src/densities/abstract_density.jl#L163-L177
and
https://github.com/bat/BAT.jl/blob/a934f6845111aba431e444fd11be32676ff5543c/src/densities/abstract_density.jl#L197-L203

**Comment by oschulz** [comment]
Ignore `_strip_duals` for now, that may change in the future.

**Comment by sthayashi** [comment]
Will check build with updated master, then ready for review.

**Comment by oschulz** [comment]
Thanks, let me know if it works!

**Comment by sthayashi** [comment]
Should I change `logvalof` and all related functions in these tests to `BAT.logvalof`, or use the new functions, e.g. `BAT.eval_logval`?

**Comment by sthayashi** [comment]
I just saw that `BAT.logvalof_unchecked` was completely removed, so I guess use the `BAT.eval_...` versions?

**Comment by oschulz** [comment]
Yes, `BAT.logvalof_unchecked` is now called `BAT.eval_logval_unchecked`.

**Comment by sthayashi** [comment]
Ready for review

**Comment by sthayashi** [comment]
Just saw the typo in the commit message. Should be `test_abstract_density.jl`.

**Review by oschulz — CHANGES_REQUESTED** [review]

File: test/densities/test_abstract_density.jl [hunk 1]
```diff
@@ -9,14 +9,94 @@ using ArraysOfArrays, ValueShapes, Distributions, PDMats, StatsBase
...
+    x_for_sd_bad_shape = incorrect_shape_of_sd(x_for_sd)
+
+    sd = _ShapeDensityStruct()
+
+    @test BAT.eval_logval(sd, x_for_sd_good_shape) == logpdf(mvn, x1_for_sd) + logpdf(mvu, x2_for_sd)
+    @test_throws ArgumentError BAT.eval_logval(sd, x_for_sd_bad_shape)
 
     @testset "rand" begin
         td = _TestDensityStruct()
         @test rand(MersenneTwister(7002), sampler(td)) ≈ [-2.415270938, 0.7070171342, 1.0224848653]
```
[the remark this answers is not in the document]
Comment (sthayashi) [reply test/densities/test_abstract_density.jl]: I didn't actually put that in there. I could add similar tests for the other test structs used if it's good to have.

[the remark this answers is not in the document]
Comment (oschulz) [reply test/densities/test_abstract_density.jl]: Yes, I didn't know about that issue back then (and StableRNG didn't exist yet). We should get rid of similar instances in the tests, but that can be a separate PR. Thanks, I'll merge this one now.

**Comment by oschulz** [comment]
Could you rebase this on the current master and update as necessary?

**Comment by sthayashi** [comment]
Tasks before ready for review
- [ ]  Clean commit history

**Comment by oschulz** [comment]
Thanks!

**Pull request merged**